src/tuvx/tuvx.cpp

```diff
@@ -14,58 +14,194 @@
+    }
+    catch (const std::system_error &e)
+    {
+      *error = ToError(e);
+    }
+  }
 
-    return (*error_code == 0) ? tuvx : nullptr;
+  GridMap *GetGridMap(TUVX *tuvx, Error *error)
+  {
+    DeleteError(error);
+    return tuvx->CreateGridMap(error);
   }
 
-  void DeleteTuvx(const TUVX *tuvx)
+  void DeleteGridMap(GridMap *grid_map, Error *error)
   {
-    delete tuvx;
+    *error = NoError();
+    try
+    {
+      delete grid_map;
+    }
+    catch (const std::system_error &e)
+    {
+      *error = ToError(e);
+    }
   }
 
-  TUVX::~TUVX()
+  Grid *GetGrid(GridMap *grid_map, const char *grid_name, const char *grid_units, Error *error)
   {
-    int error_code = 0;
-    if (tuvx_ != nullptr)
+    *error = NoError();
+    return grid_map->GetGrid(grid_name, grid_units, error);
+  }
+
+  void DeleteGrid(Grid *grid, Error *error)
+  {
+    *error = NoError();
+    try
     {
-      InternalDeleteTuvx(tuvx_.get(), &error_code);
+      delete grid;
     }
+    catch (const std::system_error &e)
+    {
+      *error = ToError(e);
+    }
+  }
+
+  void SetEdges(Grid *grid, double edges[], std::size_t num_edges, Error *error)
+  {
+    grid->SetEdges(edges, num_edges, error);
   }
 
-  int TUVX::Create(const std::string &config_path)
+  void SetMidpoints(Grid *grid, double midpoints[], std::size_t num_midpoints, Error *error)
   {
-    int parsing_status = 0;  // 0 on success, else on failure
-    String config_path_str = CreateString(const_cast<char *>(config_path.c_str()));
+    grid->SetMidpoints(midpoints, num_midpoints, error);
+  }
+
+  TUVX::TUVX()
+      : tuvx_(),
+        grid_map_(nullptr)
+  {
+  }
+
+  TUVX::~TUVX()
+  {
+    int error_code = 0;
+    if (tuvx_ != nullptr)
+      InternalDeleteTuvx(tuvx_, &error_code);
+    tuvx_ = nullptr;
+  }
 
+  void TUVX::Create(const std::string &config_path, Error *error)
+  {
+    int parsing_status = 0;  // 0 on success, 1 on failure
     try
     {
-      tuvx_ = std::make_unique<void *>(InternalCreateTuvx(config_path_str, &parsing_status));
+      String config_path_str = CreateString(const_cast<char *>(config_path.c_str()));
+      tuvx_ = InternalCreateTuvx(config_path_str, &parsing_status);
       DeleteString(&config_path_str);
+      if (parsing_status == 1)
+      {
+        *error = Error{ 1, CreateString(MUSICA_ERROR_CATEGORY), CreateString("Failed to create tuvx instance") };
+      }
+      else
+      {
+        *error = NoError();
+      }
     }
-    catch (const std::bad_alloc &e)
+    catch (const std::system_error &e)
     {
-      parsing_status = 1;
-      std::cerr << e.what() << std::endl;
-      DeleteString(&config_path_str);
+      *error = ToError(e);
     }
-    catch (const std::exception &e)
+    catch (...)
     {
-      parsing_status = 2;
-      std::cerr << e.what() << std::endl;
-      DeleteString(&config_path_str);
+      *error = Error{ 1, CreateString(MUSICA_ERROR_CATEGORY), CreateString("Failed to create tuvx instance") };
+    }
```

Comment:
well on that note all of our errors are leaking memory

Reply:
Could you elaborate what you mean? I think we should try to  prevent the memory leak as mush as possible and this leak caused by `String` is fairly obvious.

Reply:
`CreateString` allocates memory for a `char*` array. Typically we pass this over to fortran. I am not sure if we are deallocating the memory each time. It's entirely possible that we are with all of the micm stuff and I failed to replicate it with tuv. I also started another branch to check the string handling and fix it if it's wrong. If I find anything, I'll make a PR from that branch into this

Reply:
Ah okay, sounds good, thank you.

Reply:
So we don't leak memory when we are on the successful path. Everytime we pass one of these Strings to fortran, fortran calls another function `delete_string_c` that deletes the allocated memory after fortran copies it to some local memory.

However, if fortran runs into an error before it calls delete_string_c, then we will deleter memory.


I changed it to this
```
      std::unique_ptr<String, decltype(&DeleteString)> config_path_str(new String, &DeleteString);
      config_path_str->size_ = config_path.size();
      config_path_str->value_ = new char[config_path.size() + 1];
      std::strcpy(config_path_str->value_, config_path.c_str());
      tuvx_ = InternalCreateTuvx(*config_path_str, &parsing_status);
```

and we are guarnteed to always delete the string when it goes out of scope in case of an error. It won't be a double free because the `DeleteString` function checks if the `value_` of the string is still allocated. It's not thread safe though, but I don't think we need this function to be thread safe